google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,514 @@
+    # currently only relevant for ReadRows
+    application_blocking_time: float = 0.0
+    # backoff time is added to application_blocking_time
+    backoff_before_attempt: float = 0.0
+    # time waiting on grpc channel
+    # TODO: capture grpc_throttling_time
+    grpc_throttling_time: float = 0.0
+
+
+@dataclass
+class ActiveOperationMetric:
+    """
+    A dataclass representing the data associated with an rpc operation that is
+    currently in progress. Fields are mutable and may be optional.
+    """
+
+    op_type: OperationType
+    backoff_generator: Generator[float, int, None] | None = None
+    # keep both clock time and monotonic timestamps for active operations
+    start_time: TimeTuple = field(default_factory=TimeTuple)
+    active_attempt: ActiveAttemptMetric | None = None
+    cluster_id: str | None = None
+    zone: str | None = None
+    completed_attempts: list[CompletedAttemptMetric] = field(default_factory=list)
+    is_streaming: bool = False  # only True for read_rows operations
+    was_completed: bool = False
+    handlers: list[MetricsHandler] = field(default_factory=list)
+    # time waiting on flow control
+    flow_throttling_time: float = 0.0
+
+    @property
+    def state(self) -> OperationState:
+        if self.was_completed:
+            return OperationState.COMPLETED
+        elif self.active_attempt is None:
+            if self.completed_attempts:
+                return OperationState.BETWEEN_ATTEMPTS
+            else:
+                return OperationState.CREATED
+        else:
+            return OperationState.ACTIVE_ATTEMPT
+
+    def start(self) -> None:
+        """
+        Optionally called to mark the start of the operation. If not called,
+        the operation will be started at initialization.
+
+        If the operation was completed or has active attempts, will raise an
+        exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.CREATED:
+            return self._handle_error(INVALID_STATE_ERROR.format("start", self.state))
+        self.start_time = TimeTuple()
+
+    def start_attempt(self) -> None:
+        """
+        Called to initiate a new attempt for the operation.
+
+        If the operation was completed or there is already an active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if (
+            self.state != OperationState.BETWEEN_ATTEMPTS
+            and self.state != OperationState.CREATED
+        ):
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("start_attempt", self.state)
+            )
+
+        # find backoff value
+        if self.backoff_generator and len(self.completed_attempts) > 0:
+            # find the attempt's backoff by sending attempt number to generator
+            backoff = self.backoff_generator.send(len(self.completed_attempts) - 1)
```

Comment:
would this cause overhead cause we're caculating the backoff in metrics? Are there jitters on the backoff time? Would the actual backoff time between attempts be different?

Reply:
(Oops, looks like I left the implementation for the backoff generator [in the wrong PR](https://github.com/googleapis/python-bigtable/blob/20260dca391cc124c673f8f5ba0735dcb3b33feb/google/cloud/bigtable/data/_helpers.py#L101))

Python libraries share backoff generator code [from api_core](https://github.com/googleapis/python-api-core/blob/5922f2b8f382ee1fd9b69fafb2eed39db93b60fb/google/api_core/retry/retry_base.py#L119), which uses randomness and jitter

What I did for our custom generator is add a list to keep track of the values that were used. Then, we can retrieve values from the history. When we call `send(idx)` here, we're asking for the value it used for attempt `idx`. 

So to answer your questions, not much overhead, and we're getting the value used, not a new random one